components/proxy/src/main/java/com/hotels/styx/admin/handlers/ProviderRoutingHandler.java

```diff
@@ -0,0 +1,89 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.admin.handlers;
+
+import static com.hotels.styx.admin.AdminServerBuilder.adminPath;
+import static com.hotels.styx.admin.AdminServerBuilder.adminEndpointPath;
+
+import com.hotels.styx.api.Eventual;
+import com.hotels.styx.api.HttpInterceptor;
+import com.hotels.styx.api.HttpRequest;
+import com.hotels.styx.api.HttpResponse;
+import com.hotels.styx.api.WebServiceHandler;
+import com.hotels.styx.api.configuration.ObjectStore;
+import com.hotels.styx.common.http.handler.HttpStreamer;
+import com.hotels.styx.routing.db.StyxObjectStore;
+import com.hotels.styx.routing.handlers.ProviderObjectRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+
+/**
+ * Routes admin requests to the admin endpoints of each {@link com.hotels.styx.api.extension.service.spi.StyxService}
+ * in the Provider {@link ObjectStore}, and to the index page that organizes and lists these endpoints.
+ * This handler registers as a watcher on the store, and will keep the endpoint list up to date as the store data changes.
+ */
+public class ProviderRoutingHandler implements WebServiceHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ProviderRoutingHandler.class);
+    private static final int MEGABYTE = 1024 * 1024;
+    private static final String ADMIN_ROOT = "providers";
+
+    private volatile UrlPatternRouter router;
+
+    /**
+     * Create a new handler for the given provider object store.
+     * @param providerDb the provider object store
+     */
+    public ProviderRoutingHandler(StyxObjectStore<ProviderObjectRecord> providerDb) {
+        Flux.from(providerDb.watch()).subscribe(
+                this::refreshRoutes,
+                error -> LOG.error("Error in providerDB subscription", error));
+    }
+
+    @Override
+    public Eventual<HttpResponse> handle(HttpRequest request, HttpInterceptor.Context context) {
+        return router.handle(request, context);
+    }
+
+    private void refreshRoutes(ObjectStore<ProviderObjectRecord> db) {
+        LOG.info("Refreshing provider admin endpoint routes");
+        router = buildRouter(db);
+    }
+
+    private static UrlPatternRouter buildRouter(ObjectStore<ProviderObjectRecord> db) {
+        UrlPatternRouter.Builder routeBuilder = new UrlPatternRouter.Builder()
+                .get("/admin/" + ADMIN_ROOT, new ProviderListHandler(db));
+        db.entrySet().forEach(entry -> {
+                String providerName = entry.getKey();
+                entry.getValue().getStyxService().adminInterfaceHandlers(providerPath(providerName))
+                        .forEach((relPath, handler) ->
+                            routeBuilder.get(endpointPath(providerName, relPath), new HttpStreamer(MEGABYTE, handler))
```

Comment:
We could simplify this code a bit by tweaking the `UrlpatternRouter`. So if `UrlPatternRouter` took a path prefix constructor argument, in this case `/admin/providers`, then all the subsequent calls to `routeBuilder.get` could simply pass in the relative path.

I think that would simplify this code a bit and we could possibly get rid of `providerPath` and `endpointPath` methods.